[AMDGPU][SDAG] Handle ISD::PTRADD in VOP3 patterns

ritter-x2a opened the pull request:
This patch mirrors similar patterns for ISD::ADD. The main difference is
that ISD::ADD is commutative, so that a pattern definition for, e.g.,
(add (mul x, y), z), automatically also handles (add z, (mul x, y)).
ISD::PTRADD is not commutative, so we would need to handle these cases
explicitly. This patch only implements (ptradd z, (op x, y)) patterns,
where the nested operation (shift or multiply) is the offset of the
ptradd (i.e., the right operand), since base pointers that are the
result of a shift or multiply seem less likely.

For SWDEV-516125.

ritter-x2a committed:
[AMDGPU][SDAG] Handle ISD::PTRADD in VOP3 patterns

This patch mirrors similar patterns for ISD::ADD. The main difference is
that ISD::ADD is commutative, so that a pattern definition for, e.g.,
(add (mul x, y), z), automatically also handles (add z, (mul x, y)).
ISD::PTRADD is not commutative, so we would need to handle these cases
explicitly. This patch only implements (ptradd z, (op x, y)) patterns,
where the nested operation (shift or multiply) is the offset of the
ptradd (i.e., the right operand), since base pointers that are the
result of a shift or multiply seem less likely.

For SWDEV-516125.

arsenm commented on llvm/lib/Target/AMDGPU/VOP3Instructions.td:
```diff
@@ -908,19 +919,24 @@ multiclass IMAD32_Pats <VOP3_Pseudo inst> {
-    (inst $src0, $src1, $src2, 0 /* clamp */)
-    >;
+class IMAD32_Mul24_Pats_Impl<VOP3_Pseudo inst, SDPatternOperator AddOp, bit mulIsRight = 0> : GCNPat <
+    !if(mulIsRight, (i64 (AddOp i64:$src2, (i64 (AMDGPUmul_u24 i32:$src0, i32:$src1)))),
+                    (i64 (AddOp (i64 (AMDGPUmul_u24 i32:$src0, i32:$src1)), i64:$src2))),
```
Should really avoid this, commutable is supposed to be automatic. It may require a special case for ptradd in tablegen itself

ritter-x2a replied:
@arsenm, what do you think of this way to make `ptradd` commutable in some DAG patterns?

Alternatively, I could add a field to the `Pattern` tablegen class, similarly to `GISelShouldIgnore`, that is queried in `CodeGenDAGPatterns`' `GenerateVariants`, for example:
- a single `bit PtrAddCommutative` that is hardcoded to mean that `ptradd` should be treated as commutative in this pattern or
- a less hard-coded `list<SDPatternOperator> ForceOperatorCommutative` that gets a list of operators that should be treated as commutative in this pattern.

As far as I can see, that would however just add more complexity in the tablegen backend to achieve the same goal as the current or the previous implementation in this PR.

ritter-x2a replied:
Ping.

arsenm replied:
In this context the semantics of ptradd are being stripped away.

In principle you could check that the input operands are mapping into a commutable machine instruction, but that might be a pain to implement. 

The PatFrags is probably the simplest way to achieve this

ritter-x2a committed:
Use ptradd_commutable PatFrags to make ptradd commutable in some DAG patterns

arsenm commented on llvm/lib/Target/AMDGPU/VOP3Instructions.td:
```diff
@@ -9,6 +9,12 @@
 def BITOP3_32 : ComplexPattern<i32, 4, "SelectBITOP3", [and, or, xor]>;
 def BITOP3_16 : ComplexPattern<i16, 4, "SelectBITOP3", [and, or, xor]>;
 
+// Matches PTRADD as a commutative operation. Patterns using this PatFrag must
+// set GISelShouldIgnore = 1 as commuting the corresponding G_PTR_ADD is
+// invalid.
```
It's not really invalid. The semantics are being stripped away and the types will be dropped at the end of selection. The ptradd will no longer exist and no longer have meaningful type constraints

ritter-x2a replied:
Conceptually, I agree. However, enabling the GISel import for these patterns leads to crashes in callees of `testMIPredicate_MI` -- I first thought those were assertion failures, but looking closer at them now, they are actually segfaults.
My guess is that that's because GISel's `G_PTR_ADD` requires different types for its arguments (a pointer and an integer), so swapping them here breaks assumptions in the code, but I will do some more debugging to find out.

ritter-x2a replied:
I'm not sure how much effort it is to fix this properly in the GISel matcher optimizer. Since the GISel import isn't integral to what this PR wants to achieve, I'd suggest we keep the GISelShouldIgnore = 1 for the patterns using ptradd_commutative here with an updated comment with the reason for that (as in the most recent version).
The GISel import can then be fixed separately.

ritter-x2a replied:
I fixed the issue with the GISel match table optimization with #159329 and removed the `GISelShouldIgnore`s. Does this resolve your remaining concerns with this PR, @arsenm?

ritter-x2a committed:
Update the comment suggesting GISelShouldIgnore = 1 with a more accurate reason.

ritter-x2a committed:
Drop GISelShouldIgnore for the ptradd_commutative uses

The problem that made them necessary was fixed in #159329.
This improves code generation with GISel for scaled offset computations and
makes it more consistent with SDAG.
